File: Infra/Infra/Phore.cpp
#include "Phore.hpp"

CppClassNew(Phore)

Int Phore_Init(Int o)
{
    Phore* m;
    m = CP(o);
    Int initCount;
    initCount = m->InitCount;
    int ua;
    ua = initCount;
    m->Intern = new QSemaphore(ua);
    return true;
}

Int Phore_Final(Int o)
{
    Phore* m;
    m = CP(o);

    delete m->Intern;
    return true;
}

CppField(Phore, InitCount)

Int Phore_CountGet(Int o)
{
    Phore* m;
    m = CP(o);
    int u;
    u = m->Intern->available();
    Int a;
    a = u;
    return a;
}

FieldDefaultSet(Phore, Count)

Int Phore_Open(Int o)
{
    Phore* m;
    m = CP(o);
    m->Intern->acquire();
    return true;
}

Int Phore_Close(Int o)
{
    Phore* m;
    m = CP(o);
    m->Intern->release();
    return true;
}

File: Avalon/Avalon.Thread/Phore.cs
namespace Avalon.Thread;

public class Phore : Any
{
    public override bool Init()
    {
        base.Init();
        ulong ua;
        ua = (ulong)(this.InitCount);
        this.Intern = Extern.Phore_New();
        Extern.Phore_InitCountSet(this.Intern, ua);
        Extern.Phore_Init(this.Intern);
        return true;
    }

    public virtual bool Final()
    {
        Extern.Phore_Final(this.Intern);
        Extern.Phore_Delete(this.Intern);
        return true;
    }

    public virtual long InitCount { get; set; }
    public virtual long Count
    {
        get
        {
            ulong u;
            u = Extern.Phore_CountGet(this.Intern);
            long a;
            a = (long)u;
            return a;
        }
        set
        {
        }
    }
    private ulong Intern { get; set; }

    public virtual bool Open()
    {
        Extern.Phore_Open(this.Intern);
        return true;
    }

    public virtual bool Close()
    {
        Extern.Phore_Close(this.Intern);
        return true;
    }
}